src/main/java/fileio_use_case/application_business/timetable_specific_classes/TimetableGatewayInteractor.java

```diff
@@ -0,0 +1,105 @@
+     * from the JSON file.
+     * @param jsonData FileImportRequestModel, Session Type
+     * @return TimetableModel
+     */
+    public TimetableModel readFromFile(FileImportRequestModel jsonData, String sessionType) throws IOException, ParseException, java.text.ParseException, InvalidSectionsException {
+        String filePath = jsonData.getFilePath();
+        Timetable aTimetable = this.timetableGateway.readFromFile(filePath, sessionType);
+        for (Flow.Subscriber<Object> subscriber : receivers){
+            subscriber.onNext(aTimetable);
+        }
+        return timetableToTimetableModel(aTimetable);
+    }
+    /** Helper method for readFromFile */
+    private TimetableModel timetableToTimetableModel(Timetable aTimetable) {
+        List<CourseModel> allCourses = new ArrayList<>();
+        for (TimetableCourse course : aTimetable.getCourseList()) {
+            allCourses.add(timetableCourseToCourseModel(course));
+        }
+        return new TimetableModel(allCourses);
+    }
+    /** Helper method for timetableToTimetableModel */
+    private CourseModel timetableCourseToCourseModel(TimetableCourse course) {
+        List<SectionModel> allSections = new ArrayList<>();
+        for (Section section : course.getSections()) {
+            List<BlockModel> allBlocks = new ArrayList<>();
+            for (Block block : section.getBlocks()){
+                allBlocks.add(new BlockModel(block.getDay(), block.getStartTime(),
+                        block.getEndTime(), block.getRoom()));
+            }
+            allSections.add(new SectionModel(section.getCode(), section.getInstructorName(), allBlocks));
+        }
+        return new CourseModel(course.getTitle(),
+                allSections, course.getCourseSession(), course.getCourseCode(),
+                course.getBreadth());
+    }
+    /** Reads timetableRequestModel into a JSON file to be stored in
+     * src/main/saved_timetables/.
+     * @param timetableRequestModel - a request model for Timetable
+     */
+    public void timetableToFile(TimetableModel timetableRequestModel, String sessionType) throws InvalidSectionsException {
+        Timetable aTimetable = timetableModelToTimetable(timetableRequestModel, sessionType);
+        timetableGateway.timetableToFile(aTimetable);
+    }
+    /** Helper method for timetableToFile */
+    private Timetable timetableModelToTimetable(TimetableModel model, String sessionType) throws InvalidSectionsException {
+        ArrayList<TimetableCourse> allCourses = new ArrayList<>();
+        for (CourseModel course : model.getCourses()) {
+            allCourses.add(courseModelToTimetableCourse(course));
+        }
+        return new Timetable(allCourses, sessionType);
+    }
+    /** Helper method for helper method timetableModelTimetable */
+    private TimetableCourse courseModelToTimetableCourse(CourseModel course) throws InvalidSectionsException {
```

Comment:
Good extraction of methods here to avoid having an overly long method, but it's still kind of duplicate code!

Reply:
emily